### PR TITLE
refactor: marketplace proxy

### PR DESCRIPTION
**Pull request by kupermind**
- Marketplace proxy.

### CODE REVIEW

File: contracts/AgentMech.sol [hunk 1]
```diff
@@ -103,8 +95,8 @@ contract AgentMech is OlasMech {
     bytes32 public immutable domainSeparator;
     // Original chain Id
     uint256 public immutable chainId;
-    // Mech Manager address
-    address public immutable mechManager;
```
Comment (kupermind) [inline contracts/AgentMech.sol:107]: No mech manager, just a mech marketplace

File: contracts/MechMarketplace.sol [hunk 2]
```diff
@@ -23,18 +39,25 @@ struct MechDelivery {
     // Requester address
     address requester;
     // Response timeout window
-    uint32 responseTimeout;
+    uint256 responseTimeout;
+    // Payment amount
+    uint256 payment;
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:44]: Payment is recorded into the delivery struct

File: contracts/MechMarketplace.sol [hunk 3]
```diff
@@ -58,53 +83,73 @@ contract MechMarketplace is IErrorsMarketplace {
...
     address public immutable serviceRegistry;
-
+    // Wrapped native token address
+    address public immutable wrappedNativeToken;
+    // Buy back burner address
+    address public immutable buyBackBurner;
+
+    // TODO: if the fee is defined here, needs its changing function
+    // Universal mech marketplace fee (max of 10_000 == 100%)
+    uint256 public fee;
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:97]: Need to decide where and how the fee is handled

File: contracts/MechMarketplace.sol [hunk 4]
```diff
@@ -150,37 +196,109 @@ contract MechMarketplace is IErrorsMarketplace {
         );
     }
 
-    /// @dev Processes payment for request delivery.
-    /// @param deliveryMech Delivery agent mech address.
-    /// @param requestId Request id.
-    function _processPayment(address deliveryMech, uint256 requestId) internal virtual {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:156]: This is moved below as a public function and works as a pull and not a push

File: contracts/MechMarketplace.sol [hunk 5]
```diff
@@ -255,9 +373,11 @@ contract MechMarketplace is IErrorsMarketplace {
         // Record priorityMech and response timeout
         mechDelivery.priorityMech = priorityMech;
         // responseTimeout from relative time to absolute time
-        mechDelivery.responseTimeout = uint32(responseTimeout + block.timestamp);
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:258]: Why even bother with casting, we are on L2

File: contracts/MechMarketplace.sol [hunk 6]
```diff
@@ -469,9 +634,8 @@ contract MechMarketplace is IErrorsMarketplace {
             revert ZeroValue();
         }
 
-        // Check mech validity
-        bool status = IMech(mech).checkMechValidity(mech);
-        if (!status) {
+        // Check mech validity as it must be created and recorded via this marketplace
+        if (mapAgentMechFactories[mech] == address(0)) {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:638]: If the mech was not recorded by this marketplace, it must not be allowed to proceed.

File: contracts/MechMarketplace.sol [hunk 7]
```diff
@@ -541,6 +705,14 @@ contract MechMarketplace is IErrorsMarketplace {
         return mapDeliveryCounts[account];
     }
 
+    // TODO Check if needed
+    /// @dev Gets deliveries count for a specific agent mech.
+    /// @param agentMech Agent mech address.
+    /// @return Deliveries count.
+    function getAgentMechDeliveriesCount(address agentMech) external view returns (uint256) {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:712]: Strip "agent" from all the mechs?

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**